miniapps/electromagnetics/maxwell_solver.cpp

```diff
@@ -0,0 +1,673 @@
+   NegCurl_ = Curl_->ParallelAssemble();
+   // Beware this modifies the matrix stored within the Curl_ object.
+   *NegCurl_ *= -1.0;
+
+   // Build grid functions
+   e_    = new ParGridFunction(HCurlFESpace_);
+   dedt_ = new ParGridFunction(HCurlFESpace_);
+   rhs_  = new ParGridFunction(HCurlFESpace_);
+   b_    = new ParGridFunction(HDivFESpace_);
+
+   E_ = e_->ParallelProject();
+   B_ = b_->ParallelProject();
+
+   HD_  = new HypreParVector(HDivFESpace_);
+   RHS_ = new HypreParVector(HCurlFESpace_);
+
+   // Initialize dedt to zero
+   *dedt_ = 0.0;
+
+   if ( j_src_)
+   {
+      if ( myid_ == 0 && logging_ > 0 )
+      {
+         cout << "Creating Current Source" << endl;
+      }
+      jCoef_ = new VectorFunctionCoefficient(3,j_src_);
+
+      j_  = new ParGridFunction(HCurlFESpace_);
+      j_->ProjectCoefficient(*jCoef_);
+
+      jd_ = new ParLinearForm(HCurlFESpace_);
+      jd_->AddDomainIntegrator(new VectorFEDomainLFIntegrator(*jCoef_));
+      jd_->Assemble();
+   }
+
+   dtMax_ = GetMaximumTimeStep();
+}
+
+MaxwellSolver::~MaxwellSolver()
+{
+   delete epsCoef_;
+   delete muInvCoef_;
+   delete etaInvCoef_;
+   delete jCoef_;
+   delete dEdtBCCoef_;
+
+   delete E_;
+   delete B_;
+   delete HD_;
+   delete RHS_;
+
+   delete e_;
+   delete b_;
+   delete j_;
+   delete dedt_;
+   delete rhs_;
+   delete jd_;
+
+   delete Curl_;
+
+   delete M1Losses_;
+   delete M2MuInv_;
+
+   delete hDivMassMuInv_;
+   delete hCurlLosses_;
+   delete weakCurlMuInv_;
+
+   delete HCurlFESpace_;
+   delete HDivFESpace_;
+
+   map<int, ParBilinearForm*>::iterator mit1;
+   for (mit1=a1_.begin(); mit1!=a1_.end(); mit1++)
+   {
+      int i = mit1->first;
+      delete pcg_[i];
+      delete diagScale_[i];
+      delete A1_[i];
+      delete a1_[i];
+   }
+
+   map<int, Coefficient*>::iterator mit2;
+   for (mit2=dtCoef_.begin(); mit2!=dtCoef_.end(); mit2++)
+   {
+      delete mit2->second;
+   }
+   for (mit2=dtSigmaCoef_.begin(); mit2!=dtSigmaCoef_.end(); mit2++)
+   {
+      delete mit2->second;
+   }
+   for (mit2=dtEtaInvCoef_.begin(); mit2!=dtEtaInvCoef_.end(); mit2++)
+   {
+      delete mit2->second;
+   }
+
+   map<string, socketstream*>::iterator mit3;
+   for (mit3=socks_.begin(); mit3!=socks_.end(); mit3++)
+   {
+      delete mit3->second;
+   }
+}
+
+HYPRE_Int
+MaxwellSolver::GetProblemSize()
+{
+   return HCurlFESpace_->GlobalTrueVSize();
+}
+
+void
+MaxwellSolver::PrintSizes()
+{
+   HYPRE_Int size_nd = HCurlFESpace_->GlobalTrueVSize();
+   HYPRE_Int size_rt = HDivFESpace_->GlobalTrueVSize();
+   if ( myid_ == 0 )
+   {
+      cout << "Number of H(Curl) unknowns: " << size_nd << endl;
+      cout << "Number of H(Div)  unknowns: " << size_rt << endl << flush;
+   }
+}
+
+void
+MaxwellSolver::SetInitialEField(VectorCoefficient & EFieldCoef)
+{
+   eCoef_ = &EFieldCoef;
+   e_->ProjectCoefficient(EFieldCoef);
+   e_->ParallelProject(*E_);
+}
+
+void
+MaxwellSolver::SetInitialBField(VectorCoefficient & BFieldCoef)
+{
```

Comment:
@mlstowell 
I am hoping this would make more better 
  b_->ProjectCoefficient(*bCoef_);